" change this to return `FnMut` instead of `Fn`" where the return type is a std::result::Result

I tried this code:
https://github.com/ambaxter/bbolt-rs/commit/69eb00a81ec0615a2914dbc63768f5defc046e81#diff-4007187965c1fea4fd252a76f5f1007b2ea59f6d86e2f454edee547c5036be24R20

I expected to see this happen:  Code compilation, or at least a sensible error.

Instead, this happened: 

```
error[E0596]: cannot borrow `*f` as mutable, as it is a captured variable in a `Fn` closure
  --> src/util.rs:20:17
   |
14 | fn walk<F>(src: &Bolt, f: &mut F) -> crate::Result<()>
   |    ----                              ----------------- change this to return `FnMut` instead of `Fn`
...
18 |   src.view(|tx| {
   |            ---- in this closure
19 |     let tx_root = Breadcrumb::Root(b"tx".as_slice());
20 |     tx.for_each(|name, b| walk_bucket(&b, &tx_root, name, None, b.sequence(), f))
   |                 ^^^^^^^^^ cannot borrow as mutable                            - mutable borrow occurs due to use of `*f` in closure

For more information about this error, try `rustc --explain E0596`.

```

I tried to make a reproducer using the following code that expresses the same type and fn signatures on Rust Playground, but I'm unable to trigger it.

```rust
#![allow(dead_code)]
#![allow(unused_variables)]

use std::marker::PhantomData;

type Result<T, E = String> = std::result::Result<T, E>;

struct Bolt {}

trait BoltApi {
    fn view<'tx, F: FnMut(TxRef<'tx>) -> crate::Result<()>>(&'tx self, f: F) -> crate::Result<()>;
}

impl BoltApi for Bolt {
    fn view<'tx, F: FnMut(TxRef<'tx>) -> crate::Result<()>>(
        &'tx self,
        mut f: F,
    ) -> crate::Result<()> {
        Ok(())
    }
}

struct BucketImpl<'tx> {
    _tx: PhantomData<&'tx u64>,
}

trait BucketApi<'tx> {
    fn for_each<F: FnMut(&'tx [u8], Option<&'tx [u8]>) -> crate::Result<()>>(
        &self,
        f: F,
    ) -> crate::Result<()>;
}

impl<'tx> BucketApi<'tx> for BucketImpl<'tx> {
    fn for_each<F: FnMut(&'tx [u8], Option<&'tx [u8]>) -> crate::Result<()>>(
        &self,
        f: F,
    ) -> crate::Result<()> {
        Ok(())
    }
}

struct TxRef<'tx> {
    _tx: PhantomData<&'tx u64>,
}

trait TxApi<'tx> {
    fn for_each<F: FnMut(&[u8], BucketImpl<'tx>) -> crate::Result<()>>(
        &self,
        f: F,
    ) -> crate::Result<()>;
}

impl<'tx> TxApi<'tx> for TxRef<'tx> {
    fn for_each<F: FnMut(&[u8], BucketImpl<'tx>) -> crate::Result<()>>(
        &self,
        f: F,
    ) -> crate::Result<()>
where {
        Ok(())
    }
}

struct Tx {}

fn walk<F: FnMut(&[u8], Option<&[u8]>, u64) -> crate::Result<()>>(
    src: &Bolt,
    f: &mut F,
) -> crate::Result<()> {
    src.view(|tx| tx.for_each(|name, b| tx.for_each(|k, v| walk_bucket(&b, k, None, 0, f))))
}

fn walk_bucket<F: FnMut(&[u8], Option<&[u8]>, u64) -> crate::Result<()>>(
    b: &BucketImpl,
    key: &[u8],
    value: Option<&[u8]>,
    seq: u64,
    f: &mut F,
) -> crate::Result<()> {
    f(key, value, seq)?;
    b.for_each(|k, v| walk_bucket(b, k, v, 0, f))
}

fn main() {}
```


### Meta

`rustc --version --verbose`:
```
% rustc --version --verbose
rustc 1.78.0 (9b00956e5 2024-04-29)
binary: rustc
commit-hash: 9b00956e56009bab2aa15d7bff10916599e3d6d6
commit-date: 2024-04-29
host: x86_64-apple-darwin
release: 1.78.0
LLVM version: 18.1.2

```

tested on Nightly as well.

<!--
Include a backtrace in the code block by setting `RUST_BACKTRACE=1` in your
environment. E.g. `RUST_BACKTRACE=1 cargo build`.
-->
<details><summary>Backtrace</summary>
<p>

```
 % RUST_BACKTRACE=1 cargo build
   Compiling bbolt-rs v1.3.9 (/Users/abaxter/RustroverProjects/bucket_test)
warning: unused import: `crate::common::cell::RefCell`
 --> src/common/pool.rs:1:5
  |
1 | use crate::common::cell::RefCell;
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: `#[warn(unused_imports)]` on by default

warning: unused import: `std::rc::Rc`
 --> src/common/pool.rs:6:5
  |
6 | use std::rc::Rc;
  |     ^^^^^^^^^^^

warning: unused import: `size::Size`
 --> src/util.rs:3:5
  |
3 | use size::Size;
  |     ^^^^^^^^^^

warning: unused import: `DerefMut`
  --> src/bucket.rs:24:34
   |
24 | use std::ops::{AddAssign, Deref, DerefMut};
   |                                  ^^^^^^^^

warning: unused import: `std::io::Write`
  --> src/tx.rs:28:5
   |
28 | use std::io::Write;
   |     ^^^^^^^^^^^^^^

warning: unused variable: `size`
 --> src/util.rs:9:11
  |
9 |   let mut size = 0u64;
  |           ^^^^ help: if this is intentional, prefix it with an underscore: `_size`
  |
  = note: `#[warn(unused_variables)]` on by default

warning: unused variable: `dst_tx`
  --> src/util.rs:10:11
   |
10 |   let mut dst_tx = dst.begin_rw_tx()?;
   |           ^^^^^^ help: if this is intentional, prefix it with an underscore: `_dst_tx`

warning: unused variable: `max_size`
 --> src/util.rs:7:44
  |
7 | pub fn compact(src: &Bolt, dst: &mut Bolt, max_size: u64) -> crate::Result<()> {
  |                                            ^^^^^^^^ help: if this is intentional, prefix it with an underscore: `_max_size`

warning: variable does not need to be mutable
 --> src/util.rs:9:7
  |
9 |   let mut size = 0u64;
  |       ----^^^^
  |       |
  |       help: remove this `mut`
  |
  = note: `#[warn(unused_mut)]` on by default

warning: variable does not need to be mutable
  --> src/util.rs:10:7
   |
10 |   let mut dst_tx = dst.begin_rw_tx()?;
   |       ----^^^^^^
   |       |
   |       help: remove this `mut`

error[E0596]: cannot borrow `*f` as mutable, as it is a captured variable in a `Fn` closure
  --> src/util.rs:20:17
   |
14 | fn walk<F>(src: &Bolt, f: &mut F) -> crate::Result<()>
   |    ----                              ----------------- change this to return `FnMut` instead of `Fn`
...
18 |   src.view(|tx| {
   |            ---- in this closure
19 |     let tx_root = Breadcrumb::Root(b"tx".as_slice());
20 |     tx.for_each(|name, b| walk_bucket(&b, &tx_root, name, None, b.sequence(), f))
   |                 ^^^^^^^^^ cannot borrow as mutable                            - mutable borrow occurs due to use of `*f` in closure

For more information about this error, try `rustc --explain E0596`.
warning: `bbolt-rs` (lib) generated 10 warnings
error: could not compile `bbolt-rs` (lib) due to 1 previous error; 10 warnings emitted
```

</p>
</details>


<!-- TRIAGEBOT_START -->

<!-- TRIAGEBOT_ASSIGN_START -->

<!-- TRIAGEBOT_ASSIGN_DATA_START$${"user":"gurry"}$$TRIAGEBOT_ASSIGN_DATA_END -->

<!-- TRIAGEBOT_ASSIGN_END -->
<!-- TRIAGEBOT_END -->

--- a/compiler/rustc_borrowck/src/diagnostics/mutability_errors.rs
+++ b/compiler/rustc_borrowck/src/diagnostics/mutability_errors.rs
@@ -937,56 +937,81 @@ impl<'a, 'tcx> MirBorrowckCtxt<'a, 'tcx> {
         let node = self.infcx.tcx.hir_node(fn_call_id);
         let def_id = hir.enclosing_body_owner(fn_call_id);
         let mut look_at_return = true;
-        // If we can detect the expression to be an `fn` call where the closure was an argument,
-        // we point at the `fn` definition argument...
-        if let hir::Node::Expr(hir::Expr { kind: hir::ExprKind::Call(func, args), .. }) = node {
-            let arg_pos = args
+
+        // If the HIR node is a function or method call gets the def ID
+        // of the called function or method and the span and args of the call expr
+        let get_call_details = || {
+            let hir::Node::Expr(hir::Expr { hir_id, kind, .. }) = node else {
+                return None;
+            };
+
+            let typeck_results = self.infcx.tcx.typeck(def_id);
+
+            match kind {
+                hir::ExprKind::Call(expr, args) => {
+                    if let Some(ty::FnDef(def_id, _)) =
+                        typeck_results.node_type_opt(expr.hir_id).as_ref().map(|ty| ty.kind())
+                    {
+                        Some((*def_id, expr.span, *args))
+                    } else {
+                        None
+                    }
+                }
+                hir::ExprKind::MethodCall(_, _, args, span) => {
+                    if let Some(def_id) = typeck_results.type_dependent_def_id(*hir_id) {
+                        Some((def_id, *span, *args))
+                    } else {
+                        None
+                    }
+                }
+                _ => None,
+            }
+        };
+
+        // If we can detect the expression to be an function or method call where the closure was an argument,
+        // we point at the function or method definition argument...
+        if let Some((callee_def_id, call_span, call_args)) = get_call_details() {
+            let arg_pos = call_args
                 .iter()
                 .enumerate()
                 .filter(|(_, arg)| arg.hir_id == closure_id)
                 .map(|(pos, _)| pos)
                 .next();
-            let tables = self.infcx.tcx.typeck(def_id);
-            if let Some(ty::FnDef(def_id, _)) =
-                tables.node_type_opt(func.hir_id).as_ref().map(|ty| ty.kind())
-            {
-                let arg = match hir.get_if_local(*def_id) {
-                    Some(
-                        hir::Node::Item(hir::Item {
-                            ident, kind: hir::ItemKind::Fn(sig, ..), ..
+
+            let arg = match hir.get_if_local(callee_def_id) {
+                Some(
+                    hir::Node::Item(hir::Item { ident, kind: hir::ItemKind::Fn(sig, ..), .. })
+                    | hir::Node::TraitItem(hir::TraitItem {
+                        ident,
+                        kind: hir::TraitItemKind::Fn(sig, _),
+                        ..
+                    })
+                    | hir::Node::ImplItem(hir::ImplItem {
+                        ident,
+                        kind: hir::ImplItemKind::Fn(sig, _),
+                        ..
+                    }),
+                ) => Some(
+                    arg_pos
+                        .and_then(|pos| {
+                            sig.decl.inputs.get(
+                                pos + if sig.decl.implicit_self.has_implicit_self() {
+                                    1
+                                } else {
+                                    0
+                                },
+                            )
                         })
-                        | hir::Node::TraitItem(hir::TraitItem {
-                            ident,
-                            kind: hir::TraitItemKind::Fn(sig, _),
-                            ..
-                        })
-                        | hir::Node::ImplItem(hir::ImplItem {
-                            ident,
-                            kind: hir::ImplItemKind::Fn(sig, _),
-                            ..
-                        }),
-                    ) => Some(
-                        arg_pos
-                            .and_then(|pos| {
-                                sig.decl.inputs.get(
-                                    pos + if sig.decl.implicit_self.has_implicit_self() {
-                                        1
-                                    } else {
-                                        0
-                                    },
-                                )
-                            })
-                            .map(|arg| arg.span)
-                            .unwrap_or(ident.span),
-                    ),
-                    _ => None,
-                };
-                if let Some(span) = arg {
-                    err.span_label(span, "change this to accept `FnMut` instead of `Fn`");
-                    err.span_label(func.span, "expects `Fn` instead of `FnMut`");
-                    err.span_label(closure_span, "in this closure");
-                    look_at_return = false;
-                }
+                        .map(|arg| arg.span)
+                        .unwrap_or(ident.span),
+                ),
+                _ => None,
+            };
+            if let Some(span) = arg {
+                err.span_label(span, "change this to accept `FnMut` instead of `Fn`");
+                err.span_label(call_span, "expects `Fn` instead of `FnMut`");
+                err.span_label(closure_span, "in this closure");
+                look_at_return = false;
             }
         }
 

--- a/tests/ui/closures/wrong-closure-arg-suggestion-125325.rs
+++ b/tests/ui/closures/wrong-closure-arg-suggestion-125325.rs
@@ -1,0 +1,29 @@
+// Regression test for #125325
+
+// Tests that we suggest changing an `impl Fn` param
+// to `impl FnMut` when the provided closure arg
+// is trying to mutate the closure env.
+// Ensures that it works that way for both
+// functions and methods
+
+struct S;
+
+impl S {
+    fn assoc_func(&self, _f: impl Fn()) -> usize {
+        0
+    }
+}
+
+fn func(_f: impl Fn()) -> usize {
+    0
+}
+
+fn test_func(s: &S) -> usize {
+    let mut x = ();
+    s.assoc_func(|| x = ());
+    //~^ cannot assign to `x`, as it is a captured variable in a `Fn` closure
+    func(|| x = ())
+    //~^ cannot assign to `x`, as it is a captured variable in a `Fn` closure
+}
+
+fn main() {}

--- a/tests/ui/closures/wrong-closure-arg-suggestion-125325.stderr
+++ b/tests/ui/closures/wrong-closure-arg-suggestion-125325.stderr
@@ -1,0 +1,28 @@
+error[E0594]: cannot assign to `x`, as it is a captured variable in a `Fn` closure
+  --> $DIR/wrong-closure-arg-suggestion-125325.rs:23:21
+   |
+LL |     fn assoc_func(&self, _f: impl Fn()) -> usize {
+   |                              --------- change this to accept `FnMut` instead of `Fn`
+...
+LL |     s.assoc_func(|| x = ());
+   |       --------------^^^^^^-
+   |       |          |  |
+   |       |          |  cannot assign
+   |       |          in this closure
+   |       expects `Fn` instead of `FnMut`
+
+error[E0594]: cannot assign to `x`, as it is a captured variable in a `Fn` closure
+  --> $DIR/wrong-closure-arg-suggestion-125325.rs:25:13
+   |
+LL | fn func(_f: impl Fn()) -> usize {
+   |             --------- change this to accept `FnMut` instead of `Fn`
+...
+LL |     func(|| x = ())
+   |     ---- -- ^^^^^^ cannot assign
+   |     |    |
+   |     |    in this closure
+   |     expects `Fn` instead of `FnMut`
+
+error: aborting due to 2 previous errors
+
+For more information about this error, try `rustc --explain E0594`.

